src/coreclr/src/jit/emitarm64.cpp

```diff
@@ -14882,4 +14873,99 @@ emitter::insExecutionCharacteristics emitter::getInsExecutionCharacteristics(ins
 
 #endif // defined(DEBUG) || defined(LATE_DISASM)
 
+//----------------------------------------------------------------------------------------
+// IsRedundantMov:
+//    Check if the current `mov` instruction is redundant and can be omitted.
+//    A `mov` is redundant in following 3 cases:
+//
+//    1. Move to same register
+//
+//         mov Rx, Rx
+//
+//    2. Move that is identical to last instruction emitted.
+//
+//         mov Rx, Ry  # <-- last instruction
+//         mov Rx, Ry  # <-- current instruction can be omitted.
+//
+//    3. Opposite Move as that of last instruction emitted.
+//
+//         mov Rx, Ry  # <-- last instruction
+//         mov Ry, Rx  # <-- current instruction can be omitted.
+//
+// Arguments:
+//    ins - The current instruction
+//    dst - The current destination
+//    src - The current source
+//
+// Return Value:
+//    true if previous instruction moved from current dst to src.
+
+bool emitter::IsRedundantMov(instruction ins, emitAttr size, regNumber dst, regNumber src)
+{
+    assert(ins == INS_mov);
+
+    if (dst == src)
+    {
+        // A mov with a EA_4BYTE has the side-effect of clearing the upper bits
+        // So only eliminate mov instructions that are not clearing the upper bits
+        //
+        if (isGeneralRegisterOrSP(dst) && (size == EA_8BYTE))
+        {
+            JITDUMP("\n -- suppressing mov because src and dst is same 8-byte register.\n");
+            return true;
+        }
+        else if (isVectorRegister(dst) && (size == EA_16BYTE))
+        {
+            JITDUMP("\n -- suppressing mov because src and dst is same 16-byte register.\n");
+            return true;
+        }
+    }
+
+    bool isFirstInstrInBlock = (emitCurIGinsCnt == 0) && ((emitCurIG->igFlags & IGF_EXTEND) == 0);
+
+    if (!isFirstInstrInBlock && // Don't optimize if instruction is not the first instruction in IG.
+        (emitLastIns != nullptr) &&
+        (emitLastIns->idIns() == INS_mov) && // Don't optimize if last instruction was not 'mov'.
+        (emitLastIns->idOpSize() == size))   // Don't optimize if operand size is different than previous instruction.
+    {
+        // Check if we did same move in prev instruction except dst/src were switched.
+        regNumber prevDst = emitLastIns->idReg1();
+        regNumber prevSrc = emitLastIns->idReg2();
+
+        if ((prevDst == dst) && (prevSrc == src))
+        {
+            assert(emitLastIns->idOpSize() == size);
+            JITDUMP("\n -- suppressing mov because previous instruction already moved from src to dst register.\n");
+            return true;
+        }
+
+        if ((prevDst == src) && (prevSrc == dst))
```

Comment:
How are the various vector moves handled? You wouldn't want to optimize `mov Vd[],Vn[]` for example, and that seems like it can only be determined by checking `IF`. @echesakovMSFT to confirm? I guess it's because those, plus the general-purpose `mov Rd,imm` don't go through `emitIns_R_R`, they go through `emitIns_R_R_I` (for example).

It seems like we shouldn't eliminate moves between vector/fp and general-purpose register files, which I think you've prevented here.

But you do allow:
```
mov d0, d1
mov d1,d0 // 64-bit; can eliminate
```

Is that because in the FP case (as opposed to the `mov w0, w1` general-purpose register case), the upper 64 bits are not zeroed? Looks to me like the upper bits get zeroed and this shouldn't be optimized.

Reply:
`mov Vd[], Vn[]` is done using `emitIns_R_R_I_I()`. For FP, that's a good point. @echesakovMSFT - do you know if we use `emitIns_R_R()` for it? If yes, how can I differentiate? Here is how we check for floating point register today:
```cpp
inline static bool isFloatReg(regNumber reg)
{
    return isVectorRegister(reg);
}
```

Reply:
I believe you need to consider prevSize, so if you had a case like this
```asm
mov Dd, Dn
mov Qn, Qd // upper 64-bit part of Qn could be different from Qd
```
you wouldn't eliminate the second instruction.

You need to make sure that `prevSize >= size`, otherwise, forbid elimination

Reply:
> For FP, that's a good point. @echesakovMSFT - do you know if we use emitIns_R_R() for it? If yes, how can I differentiate? 

I don't think I understand your question - how we can differentiate between what exactly?

Reply:
@kunalspathak Checking the register is not sufficient since for SIMD/FP, lanes and other things are represented outside of the register name. But as I pointed out, those various things go through different code paths.

Reply:
> Looks to me like the upper bits get zeroed and this shouldn't be optimized.

@BruceForstall - Yes. The FP `fmov` happens with `IF_DV_2G` which is different code path.

> You need to make sure that prevSize >= size, otherwise, forbid elimination

@echesakovMSFT  - I already have a check of `(emitLastIns->idOpSize() == size)`.